File: web/src/grouping.ts
import type { GroupingSpec, MedianRecord } from "./types";

export type GroupingInput = "median" | number | GroupingSpec;

export function normalizeGrouping(input: GroupingInput): GroupingSpec {
  const spec: GroupingSpec =
    input === "median"
      ? { kind: "median" }
      : typeof input === "number"
        ? { kind: "tpm", threshold: input }
        : input;
  if (!spec || typeof spec !== "object")
    throw new Error("Choose a valid comparison.");
  const keys: Record<string, string[]> = {
    median: ["kind"],
    mean: ["kind"],
    tpm: ["kind", "threshold"],
    percentile: ["kind", "percentile"],
    extremes: ["kind", "lowerPercent", "upperPercent"],
  };
  if (
    !keys[spec.kind] ||
    Object.keys(spec).some((key) => !keys[spec.kind].includes(key))
  ) {
    throw new Error(
      "Grouping parameters do not match the selected comparison.",
    );
  }
  if (
    spec.kind === "tpm" &&
    (!Number.isFinite(spec.threshold) || spec.threshold < 0)
  ) {
    throw new Error("TPM threshold must be a finite, non-negative number.");
  }
  if (spec.kind === "percentile") {
    if (
      !Number.isFinite(spec.percentile) ||
      !(spec.percentile > 0 && spec.percentile < 100)
    ) {
      throw new Error("Percentile must be greater than 0 and less than 100.");
    }
    if (spec.percentile === 50) return { kind: "median" };
  }
  if (spec.kind === "extremes") {
    const { lowerPercent: lo, upperPercent: hi } = spec;
    if (
      !Number.isFinite(lo) ||
      !Number.isFinite(hi) ||
      !(lo > 0 && hi > 0 && lo < 100 && hi < 100 && lo + hi <= 100)
    ) {
      throw new Error(
        "Group percentages must be positive and total at most 100.",
      );
    }
    if (lo + hi === 100)
      return normalizeGrouping({ kind: "percentile", percentile: lo });
  }
  return spec;
}

export function groupingLabel(spec: GroupingSpec): string {
  if (spec.kind === "median") return "Median expression";
  if (spec.kind === "mean") return "Mean (average) expression";
  if (spec.kind === "tpm") return `Expression threshold: ${spec.threshold} TPM`;
  if (spec.kind === "percentile")
    return `${spec.percentile}th percentile split`;
  return `Lowest ${Number(spec.lowerPercent.toPrecision(4))}% vs highest ${Number(spec.upperPercent.toPrecision(4))}%`;
}

export function quantile(values: number[], probability: number): number {
  if (!values.length) return Number.NaN;
  const ordered = [...values].sort((a, b) => a - b);
  // Match R type-7's one-based index and weighted-average operation order.
  const position = 1 + (ordered.length - 1) * probability;
  const lower = Math.floor(position),
    upper = Math.ceil(position),
    fraction = position - lower;
  const a = ordered[lower - 1],
    b = ordered[upper - 1];
  return fraction === 0 || a === b ? a : (1 - fraction) * a + fraction * b;
}

export function assignGroups(
  tpm: number[],
  indices: number[],
  median: MedianRecord,
  spec: GroupingSpec,
) {
  let lower = spec.kind === "tpm" ? spec.threshold : Number.NaN;
  let upper = lower;
  if (!tpm.length) return { low: [], high: [], lower, upper };
  if (spec.kind === "median") {
    lower = median.cutoff_tpm ?? quantile(tpm, 0.5);
    const flips = new Set(median.flips);
    const high = tpm.map((x, i) => x > lower !== flips.has(indices[i]));
    return { low: high.map((x) => !x), high, lower, upper: lower };
  }
  if (spec.kind === "mean") {
    let total = 0,
      correction = 0;
    for (const value of tpm) {
      const updated = total + value;
      correction +=
        Math.abs(total) >= Math.abs(value)
          ? total - updated + value
          : value - updated + total;
      total = updated;
    }
    lower = upper = (total + correction) / tpm.length;
  } else if (spec.kind === "percentile") {
    lower = upper = quantile(tpm, spec.percentile / 100);
  } else if (spec.kind === "extremes") {
    lower = quantile(tpm, spec.lowerPercent / 100);
    upper = quantile(tpm, 1 - spec.upperPercent / 100);
  }
  return {
    low: tpm.map((x) => x <= lower),
    high: tpm.map((x) => x > upper),
    lower,
    upper,
  };
}

export function groupingFromQuery(params: URLSearchParams): GroupingSpec {
  const kind = params.get("grouping");
  const number = (key: string) => {
    const value = params.get(key);
    if (
      value === null ||
      value.trim() === "" ||
      !Number.isFinite(Number(value))
    ) {
      throw new Error(`A valid ${key} is required.`);
    }
    return Number(value);
  };
  if (kind === "mean") return { kind };
  if (kind === "percentile")
    return normalizeGrouping({ kind, percentile: number("percentile") });
  if (kind === "extremes")
    return normalizeGrouping({
      kind,
      lowerPercent: number("lower"),
      upperPercent: number("upper"),
    });
  if (kind && kind !== "median")
    throw new Error("Unrecognized comparison in the link.");
  return normalizeGrouping(
    !params.has("cutoff") || params.get("cutoff") === "median"
      ? "median"
      : number("cutoff"),
  );
}

export function groupingQuery(spec: GroupingSpec): Record<string, string> {
  if (spec.kind === "median") return { cutoff: "median" };
  if (spec.kind === "tpm") return { cutoff: String(spec.threshold) };
  if (spec.kind === "mean") return { grouping: "mean" };
  if (spec.kind === "percentile")
    return { grouping: spec.kind, percentile: String(spec.percentile) };
  return {
    grouping: spec.kind,
    lower: String(spec.lowerPercent),
    upper: String(spec.upperPercent),
  };
}

File: src/survscope/grouping.py
"""Outcome-independent expression comparisons shared with the browser."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import numpy as np


@dataclass(frozen=True)
class GroupingSpec:
    """Choose two expression groups; percentages refer to endpoint-valid patients."""

    kind: str = "median"
    threshold: float | None = None
    percentile: float | None = None
    lower_percent: float | None = None
    upper_percent: float | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            key: value
            for key, value in {
                "kind": self.kind,
                "threshold": self.threshold,
                "percentile": self.percentile,
                "lowerPercent": self.lower_percent,
                "upperPercent": self.upper_percent,
            }.items()
            if value is not None
        }


def normalize_grouping(
    cutoff: str | float = "median",
    grouping: GroupingSpec | dict | None = None,
) -> GroupingSpec:
    if grouping is not None:
        if str(cutoff).lower() != "median":
            raise ValueError("Choose either grouping or a custom cutoff, not both.")
        if isinstance(grouping, dict):
            allowed = {"kind", "threshold", "percentile", "lowerPercent", "upperPercent"}
            if set(grouping) - allowed:
                raise ValueError("Unrecognized grouping parameter.")
            grouping = GroupingSpec(
                kind=grouping.get("kind", "median"),
                threshold=grouping.get("threshold"),
                percentile=grouping.get("percentile"),
                lower_percent=grouping.get("lowerPercent"),
                upper_percent=grouping.get("upperPercent"),
            )
    elif isinstance(cutoff, str) and cutoff.lower() == "median":
        grouping = GroupingSpec()
    else:
        try:
            grouping = GroupingSpec("tpm", threshold=float(cutoff))
        except (ValueError, TypeError) as error:
            raise ValueError("cutoff must be 'median' or a non-negative TPM value") from error
    if not isinstance(grouping, GroupingSpec):
        raise ValueError("grouping must be a GroupingSpec or grouping dictionary")

    def number(value: Any, name: str) -> float:
        if isinstance(value, bool) or not isinstance(value, (int, float)) or not np.isfinite(value):
            raise ValueError(f"{name} must be a finite number.")
        return float(value)

    spec = grouping
    if spec.kind not in {"median", "mean", "tpm", "percentile", "extremes"}:
        raise ValueError(f"Unsupported grouping: {spec.kind}")
    expected = {
        "median": set(),
        "mean": set(),
        "tpm": {"threshold"},
        "percentile": {"percentile"},
        "extremes": {"lowerPercent", "upperPercent"},
    }[spec.kind]
    if set(spec.to_dict()) - {"kind"} != expected:
        raise ValueError("Grouping parameters do not match the selected comparison.")
    if spec.kind == "tpm":
        if number(spec.threshold, "TPM threshold") < 0:
            raise ValueError("TPM threshold must be non-negative.")
    elif spec.kind == "percentile":
        if not 0 < number(spec.percentile, "Percentile") < 100:
            raise ValueError("Percentile must be greater than 0 and less than 100.")
        if spec.percentile == 50:
            return GroupingSpec()
    elif spec.kind == "extremes":
        lower = number(spec.lower_percent, "Lower percentage")
        upper = number(spec.upper_percent, "Upper percentage")
        if not (0 < lower < 100 and 0 < upper < 100 and lower + upper <= 100):
            raise ValueError("Group percentages must be positive and total at most 100.")
        if lower + upper == 100:
            return normalize_grouping(grouping=GroupingSpec("percentile", percentile=lower))
    elif spec.kind not in {"median", "mean"}:
        raise ValueError(f"Unsupported grouping: {spec.kind}")
    return spec


def grouping_label(spec: GroupingSpec) -> str:
    if spec.kind == "median":
        return "Median expression"
    if spec.kind == "mean":
        return "Mean (average) expression"
    if spec.kind == "tpm":
        return f"Expression threshold: {spec.threshold:g} TPM"
    if spec.kind == "percentile":
        return f"{spec.percentile:g}th percentile split"
    return f"Lowest {spec.lower_percent:.4g}% vs highest {spec.upper_percent:.4g}%"


def expression_quantile(values: np.ndarray, probability: float) -> float:
    """R type-7 operation order, including its one-based index arithmetic.

    Equivalent interpolation formulas can round to opposite sides of an
    observed value. Matching this order also preserves threshold membership.
    """
    ordered = sorted(float(value) for value in values)
    index = 1 + (len(ordered) - 1) * probability
    lo, hi = int(np.floor(index)), int(np.ceil(index))
    lower, upper = ordered[lo - 1], ordered[hi - 1]
    if index == lo or lower == upper:
        return lower
    fraction = index - lo
    return (1 - fraction) * lower + fraction * upper


def assign_groups(
    tpm: np.ndarray,
    indices: np.ndarray,
    median: dict,
    spec: GroupingSpec,
) -> tuple[np.ndarray, np.ndarray, float, float]:
    """Return low/high masks and thresholds; neither mask contains the excluded middle."""
    if len(tpm) == 0:
        threshold = float(spec.threshold) if spec.kind == "tpm" else np.nan
        return np.zeros(0, bool), np.zeros(0, bool), threshold, threshold
    if spec.kind == "median":
        stored = median.get("cutoff_tpm")
        lower = float(stored) if stored is not None else expression_quantile(tpm, 0.5)
        high = tpm > lower
        flips = set(median.get("flips", []))
        high ^= np.array([int(index) in flips for index in indices])
        return ~high, high, lower, lower
    if spec.kind == "tpm":
        lower = upper = float(spec.threshold)
    elif spec.kind == "mean":
        # Neumaier summation, with the same operation order in JavaScript.
        total = correction = 0.0
        for value in tpm:
            value = float(value)
            updated = total + value
            correction += (
                (total - updated) + value if abs(total) >= abs(value) else (value - updated) + total
            )
            total = updated
        lower = upper = (total + correction) / len(tpm)
    elif spec.kind == "percentile":
        lower = upper = expression_quantile(tpm, spec.percentile / 100)
    else:
        lower = expression_quantile(tpm, spec.lower_percent / 100)
        upper = expression_quantile(tpm, 1 - spec.upper_percent / 100)
    return tpm <= lower, tpm > upper, float(lower), float(upper)
